Add reference to entropy implementation used

mbauwens opened the pull request:
This is in reference to issue #2961.

This is just a simple addition to the [docs](https://www.nltk.org/api/nltk.lm.api.html#nltk.lm.api.LanguageModel.entropy), making it more clear that the entropy implementation in NLTK is the one based on the Shannon-McMillan-Breiman theorem, as used and referenced by Jurafsky ([Chapter 3 page 23 - 3.49](https://web.stanford.edu/~jurafsky/slp3/3.pdf)) and Jordan Boyd-Graber (`lm.txt` file in [this comment](https://github.com/nltk/nltk/issues/1342#issuecomment-203060085)).

This PR does not include the alternative implementation of the entropy function that was proposed in the issue.

mbauwens committed:
Add reference to entropy implementation used

Making it more clear that the entropy implementation in NLTK is the one based on the Shannon-McMillan-Breiman theorem, as used and referenced by Jurafsky and Jordan Boyd-Graber.

iliakur commented on nltk/lm/api.py:
```diff
@@ -162,6 +162,8 @@ def context_counts(self, context):
     def entropy(self, text_ngrams):
         """Calculate cross-entropy of model for given evaluation text.
+        This implementation is based on the Shannon-McMillan-Breiman theorem,
+        as used and referenced by Jurafsky and Jordan Boyd-Graber.
```
```suggestion
        as used and referenced by Dan Jurafsky and Jordan Boyd-Graber.
```
Consistently use full names

iliakur commented on nltk/lm/api.py:
```diff
@@ -162,6 +162,8 @@ def context_counts(self, context):
 
     def entropy(self, text_ngrams):
         """Calculate cross-entropy of model for given evaluation text.
+        This implementation is based on the Shannon-McMillan-Breiman theorem,
```
```suggestion

        This implementation is based on the Shannon-McMillan-Breiman theorem,
```
Just to stay consistent with other docstrings in the module

iliakur reviewed (changes requested):
Minor nits, otherwise lgtm, thanks!

mbauwens committed:
Consistently use full names

Co-authored-by: Ilia Kurenkov <ilia.kurenkov@gmail.com>

mbauwens committed:
Consistency with other docstrings in the module

Co-authored-by: Ilia Kurenkov <ilia.kurenkov@gmail.com>

mbauwens reviewed (commented):
Minor changes in formatting of text. Look's good!

mbauwens committed:
first draft of extended entropy method

add entropy_extended with length normalisation and relative frequency weighting. Also adapt perplexity_extended to work with this new entropy measure.

iliakur commented on nltk/lm/api.py:
```diff
@@ -179,6 +183,41 @@ def perplexity(self, text_ngrams):
         """
         return pow(2.0, self.entropy(text_ngrams))
 
+    def entropy_extended(self, text_ngrams, text_fdist, length_normalisation=True, rel_freq_weighting=False):
```
Oh, hmm, I think I'll have to review this properly I'm afraid.

One thing that worries me is lack of tests. Let's maybe do it in a separate PR, I can help you with that.

mbauwens replied:
@iliakur oh, my bad! I did not intend to already add these changes to the PR. I'm new to making PRs and didn't realise an updated fork would also be reflected into the previous PR. Will remove this from the PR!

mbauwens committed:
Added unk flag and rel freq weighting

mbauwens committed:
Cleaning up draft

mbauwens committed:
remove draft from PR

mbauwens committed:
Update api.py

mbauwens committed:
Update api.py

mbauwens commented:
@iliakur yeah, so a bit messy here, but I've reverted the PR to just the update to the docs. I now have a separate branch on my fork that I'll be getting ready for a PR for an alternative entropy/perplexity method, including tests :)

iliakur commented:
@stevenbird @tomaarsen this PR is ready to merge. I've been away from nltk development long enough that I figured I'd check in with you two about any formalities that may be missing.

tomaarsen commented:
I'm afraid I've also not had much time for NLTK recently. That said, I think this is good to go. Thanks for taking care of this, and nice to see you here again @mbauwens.

- Tom Aarsen
